File: leetcode/738.js
/**
 * @param {number} N
 * @return {number}
 */
var monotoneIncreasingDigits = function(N) {
    if (N < 10) return N;
    let last = 1 << 30, mx = 1, lx = 1, n = N;
    while (N > 0) {
        let cur = N % 10;
        if (cur > last) {
            cur = cur - 1;
            lx = mx;
        }

        mx = mx * 10;
        last = cur;
        N = parseInt(N / 10);
    }

    return (parseInt(n / lx)) * lx - (lx > 1 ? 1 : 0);
};

let res = monotoneIncreasingDigits(10);
console.log(res);

File: archive/leetcode-go/archive/738.cpp
#include <iostream>
#include "utils.h"
#include "leetcode.h"

using namespace std;

class Solution {
public:
    int monotoneIncreasingDigits(int N) {
        if (N < 10) return N;
        int rem = 0, last = INT_MAX, mx = 1;
        while (N > 0) {
            int cur = N % 10;
            if (cur > last) {
                rem = mx - 1;
                cur -= 1;
            }

            rem += cur * mx;
            mx *= 10;
            N /= 10;
            last = cur;
        }
        return rem;
    }
};

int main() {
    Solution s;
    int n;
    while (cin >> n)
        cout << s.monotoneIncreasingDigits(n) << endl;
    return 0;
}
